File: android/farmerchat-sdk/src/main/java/com/farmerchat/sdk/FarmerChatSdk.kt
package com.farmerchat.sdk

import android.content.Context
import android.provider.Settings
import android.util.Log
import com.farmerchat.sdk.activity.FarmerChatActivity
import com.farmerchat.sdk.base.AuthInitException
import com.farmerchat.sdk.di.SdkKoinHolder
import com.farmerchat.sdk.network.GuestApiClient
import com.farmerchat.sdk.preference.SdkPreferenceManager
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext

private const val TAG = "FarmerChatSdk"

object FarmerChatSdk {

    private var _config: FarmerChatConfig? = null
    private var appContext: Context? = null
    private val sdkScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    private val initMutex = Mutex()

    internal val config: FarmerChatConfig
        get() = _config ?: error("FarmerChatSdk not initialized.")

    private var initialized = false

    fun initialize(context: Context, config: FarmerChatConfig) {
        require(isValidSdkApiKey(config.sdkApiKey)) {
            "Invalid SDK API key '${config.sdkApiKey}'. " +
            "Keys must start with 'fc_live_' or 'fc_test_' followed by at least 16 alphanumeric characters. " +
            "Obtain your key from the FarmerChat developer portal."
        }
        _config = config
        appContext = context.applicationContext
        val deviceId = Settings.Secure.getString(appContext!!.contentResolver, Settings.Secure.ANDROID_ID)
        val prefs = SdkPreferenceManager(appContext!!)

        prefs.saveTokens(deviceId = deviceId)
        SdkKoinHolder.start(appContext!!)
        initialized = true

        Log.d(TAG, "initialize() — deviceId=$deviceId hasTokens=${prefs.hasTokens()}")

        if (!prefs.hasTokens()) {
            sdkScope.launch {
                try { fetchGuestTokens(prefs, deviceId, config.baseUrl) }
                catch (e: Exception) { Log.w(TAG, "Background init failed (will retry on openChat): ${e.message}") }
            }
        }
    }

    fun updateTokens(context: Context, accessToken: String, refreshToken: String) {
        SdkPreferenceManager(context.applicationContext).saveTokens(
            accessToken = accessToken,
            refreshToken = refreshToken
        )
    }

    fun openChat(context: Context, conversationId: String? = null) {
        check(initialized) { "FarmerChatSdk not initialized." }
        val ctx = context.applicationContext
        val prefs = SdkPreferenceManager(ctx)
        val targetConversationId = conversationId ?: config.conversationId

        if (prefs.hasTokens()) {
            FarmerChatActivity.launch(context, targetConversationId)
        } else {
            // Always open the chat. If init fails, the chat screen will show an error via
            // ensureTokensInternal() and let the user retry.
            sdkScope.launch {
                val deviceId = prefs.getDeviceId()
                    ?: Settings.Secure.getString(ctx.contentResolver, Settings.Secure.ANDROID_ID)
                try { fetchGuestTokens(prefs, deviceId, config.baseUrl) }
                catch (e: Exception) { Log.w(TAG, "Pre-launch init failed (chat will show error): ${e.message}") }
                withContext(Dispatchers.Main) {
                    FarmerChatActivity.launch(context, targetConversationId)
                }
            }
        }
    }

    /**
     * Ensures tokens (including user_id) are saved before opening the chat overlay.
     * Throws [AuthInitException] if initialization fails so the ViewModel can show a
     * meaningful error to the user rather than a cryptic API failure.
     */
    internal suspend fun ensureTokensInternal() {
        val ctx = appContext ?: return
        val prefs = SdkPreferenceManager(ctx)
        Log.d(TAG, "ensureTokensInternal() — hasTokens=${prefs.hasTokens()} userId=${prefs.getUserId()}")
        if (prefs.hasTokens()) return

        initMutex.withLock {
            if (prefs.hasTokens()) return
            val deviceId = prefs.getDeviceId()
                ?: Settings.Secure.getString(ctx.contentResolver, Settings.Secure.ANDROID_ID)
            try {
                fetchGuestTokens(prefs, deviceId, config.baseUrl)
            } catch (e: Exception) {
                Log.e(TAG, "Guest init failed: ${e.message}", e)
                throw AuthInitException(buildAuthErrorMessage(e.message))
            }
        }
    }

    /**
     * Calls initialize_user and saves tokens. Throws IOException on failure.
     * Callers are responsible for wrapping with user-friendly error handling.
     */
    private suspend fun fetchGuestTokens(
        prefs: SdkPreferenceManager,
        deviceId: String,
        baseUrl: String
    ) {
        withContext(Dispatchers.IO) {
            val response = GuestApiClient.initializeUser(baseUrl, deviceId)
            if (response.user_id.isNullOrEmpty()) {
                Log.e(TAG, "initialize_user returned null/empty user_id!")
                throw java.io.IOException("initialize_user returned empty user_id")
            }
            prefs.saveTokens(
                accessToken = response.access_token,
                refreshToken = response.refresh_token,
                userId = response.user_id
            )
            Log.d(TAG, "Tokens saved — userId=${response.user_id} createdNow=${response.created_now}")
        }
    }

    private fun buildAuthErrorMessage(rawMessage: String?): String {
        val msg = rawMessage ?: ""
        return when {
            // HTTP 4xx + device/limit keywords → permanent device limit error
            msg.contains("HTTP 4", ignoreCase = false) &&
                    (msg.contains("device", ignoreCase = true) || msg.contains("limit", ignoreCase = true)) ->
                "This device has reached its guest account limit. Please contact support."
            // Any other HTTP 4xx → backend rejected the request
            msg.contains("HTTP 4", ignoreCase = false) ->
                "Unable to start a chat session. Please contact support if this persists."
            // Fallback → likely a network/timeout error
            else ->
                "Please check your internet connection and try again."
        }
    }

    fun isInitialized(): Boolean = initialized

    fun shutdown() {
        SdkKoinHolder.stop()
        _config = null
        appContext = null
        initialized = false
    }

    fun clearSession(context: Context) {
        SdkPreferenceManager(context.applicationContext).clear()
    }
}

/**
 * Validates the SDK API key format.
 * Valid keys: fc_live_<16+ alphanumeric> or fc_test_<16+ alphanumeric>
 */
internal fun isValidSdkApiKey(key: String): Boolean {
    val prefix = when {
        key.startsWith("fc_live_") -> "fc_live_"
        key.startsWith("fc_test_") -> "fc_test_"
        else -> return false
    }
    val suffix = key.removePrefix(prefix)
    return suffix.length >= 16 && suffix.all { it.isLetterOrDigit() }
}

File: ios/Sources/FarmerChatSDK/FarmerChatSDK.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

// MARK: - FarmerChatConfig

/// Immutable configuration object passed to the SDK at startup.
///
/// - Parameter sdkApiKey: Your SDK API key from the FarmerChat developer portal
///   (format: `fc_live_xxx` or `fc_test_xxx`). Sent as `X-SDK-Key` on every API request.
public struct FarmerChatConfig {
    public let sdkApiKey: String
    public let baseUrl: String
    public let contentProviderId: String?
    public let conversationId: String?

    public init(
        sdkApiKey: String,
        baseUrl: String,
        contentProviderId: String? = nil,
        conversationId: String? = nil
    ) {
        self.sdkApiKey = sdkApiKey
        self.baseUrl = baseUrl
        self.contentProviderId = contentProviderId
        self.conversationId = conversationId
    }
}

// MARK: - SDKError

public enum SDKError: LocalizedError {
    case notConfigured
    case invalidBaseUrl(String)
    case invalidApiKey(String)

    public var errorDescription: String? {
        switch self {
        case .notConfigured:
            return "FarmerChatSDK has not been configured. Call FarmerChatSDK.shared.configure(_:) first."
        case .invalidBaseUrl(let url):
            return "The base URL '\(url)' is not a valid HTTP/HTTPS URL."
        case .invalidApiKey(let key):
            return "Invalid SDK API key '\(key.prefix(12))...'. " +
                   "Keys must start with 'fc_live_' or 'fc_test_' followed by at least 16 alphanumeric characters. " +
                   "Obtain your key from the FarmerChat developer portal."
        }
    }
}

// MARK: - FarmerChatSDK

/// Singleton entry point for the FarmerChat SDK.
@MainActor
public final class FarmerChatSDK {

    // MARK: Public

    public static let shared = FarmerChatSDK()

    /// The active configuration, available after `configure(_:)` is called.
    public private(set) var configuration: FarmerChatConfig?

    // MARK: Internal helpers (visible within SDK)

    internal let tokenStore: KeychainTokenStore = KeychainTokenStore()

    // MARK: Init

    private init() {}

    // MARK: Public Methods

    /// Configure the SDK. Must be called before presenting any SDK UI (e.g. in App.init or AppDelegate).
    ///
    /// Tokens are obtained automatically from the backend using the device's `identifierForVendor`.
    /// The network call happens once per fresh install; subsequent launches use the stored tokens.
    ///
    /// - Parameter config: The configuration to apply. Only `baseUrl` is required.
    /// - Throws: `SDKError.invalidBaseUrl` if the URL is malformed.
    public func configure(_ config: FarmerChatConfig) throws {
        // Validate SDK API key format: fc_live_<16+ alphanum> or fc_test_<16+ alphanum>
        guard Self.isValidApiKey(config.sdkApiKey) else {
            throw SDKError.invalidApiKey(config.sdkApiKey)
        }

        let trimmed = config.baseUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty,
              let url = URL(string: trimmed),
              let scheme = url.scheme,
              scheme == "http" || scheme == "https" else {
            throw SDKError.invalidBaseUrl(config.baseUrl)
        }

        configuration = config

        // Use a stable device ID. Prefer what's already saved (survives configure() being
        // called more than once). Fall back to identifierForVendor, then a one-time UUID.
        let deviceId: String
        if let saved = tokenStore.getDeviceId() {
            deviceId = saved
        } else {
            #if canImport(UIKit)
            deviceId = UIDevice.current.identifierForVendor?.uuidString ?? UUID().uuidString
            #else
            deviceId = UUID().uuidString
            #endif
            tokenStore.saveDeviceId(deviceId)
        }

        // Skip the network call if tokens from a previous session are already in Keychain
        guard !tokenStore.hasTokens() else { return }

        Task.detached(priority: .userInitiated) { [tokenStore] in
            do {
                let response = try await GuestAPIClient.initializeUser(
                    baseUrl: config.baseUrl,
                    deviceId: deviceId
                )
                tokenStore.saveTokens(
                    accessToken: response.access_token,
                    refreshToken: response.refresh_token
                )
                if let userId = response.user_id {
                    tokenStore.saveUserId(userId)
                }
            } catch {
                // Tokens will be fetched lazily by ensureTokens() when chat is first opened
            }
        }
    }

    /// Ensures tokens are available, fetching them via initialize_user if needed.
    /// Throws an error with a user-facing message if initialisation fails, so callers
    /// (ChatViewModel) can show a meaningful error instead of a cryptic API failure.
    internal func ensureTokens() async throws {
        guard let config = configuration else { return }
        guard !tokenStore.hasTokens() else { return }

        guard let deviceId = tokenStore.getDeviceId() else {
            throw NSError(domain: "FarmerChatSDK", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "Unable to start a chat session. Please try again."])
        }

        do {
            let response = try await GuestAPIClient.initializeUser(
                baseUrl: config.baseUrl,
                deviceId: deviceId
            )
            tokenStore.saveTokens(
                accessToken: response.access_token,
                refreshToken: response.refresh_token
            )
            if let userId = response.user_id {
                tokenStore.saveUserId(userId)
            }
        } catch {
            // Re-check: a concurrent call may have already saved tokens
            if tokenStore.hasTokens() { return }
            throw NSError(domain: "FarmerChatSDK", code: -2,
                          userInfo: [NSLocalizedDescriptionKey: buildAuthErrorMessage(error)])
        }
    }

    private func buildAuthErrorMessage(_ error: Error) -> String {
        let msg = error.localizedDescription
        if msg.contains("401") || msg.contains("403") || msg.contains("device") || msg.contains("limit") {
            return "This device has reached its guest account limit. Please contact support."
        }
        if msg.contains("400") || msg.contains("4") {
            return "Unable to start a chat session. Please contact support if this persists."
        }
        return "Please check your internet connection and try again."
    }

    // MARK: Internal Accessors

    /// Resolved base URL from active configuration.
    internal var resolvedBaseUrl: String {
        configuration?.baseUrl ?? ""
    }

    // MARK: Reset

    /// Clear all stored tokens and user data. Forces a fresh initialize_user call
    /// on the next configure (or next chat open).
    public func clearSession() {
        tokenStore.clear()
    }

    // MARK: Key Validation

    /// Returns true if the key matches: fc_live_<16+ alphanum> or fc_test_<16+ alphanum>
    static func isValidApiKey(_ key: String) -> Bool {
        let prefix: String
        if key.hasPrefix("fc_live_") { prefix = "fc_live_" }
        else if key.hasPrefix("fc_test_") { prefix = "fc_test_" }
        else { return false }

        let suffix = key.dropFirst(prefix.count)
        guard suffix.count >= 16 else { return false }
        return suffix.allSatisfy { $0.isLetter || $0.isNumber }
    }
}
